resources/mediaplayer/src/main/java/org/mobicents/media/server/impl/resource/mediaplayer/audio/CachedRemoteStreamProvider.java

```diff
@@ -0,0 +1,79 @@
+package org.mobicents.media.server.impl.resource.mediaplayer.audio;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.apache.commons.io.IOUtils;
+import org.ehcache.Cache;
+import org.ehcache.CacheManager;
+import org.ehcache.config.builders.CacheConfigurationBuilder;
+import org.ehcache.config.builders.CacheManagerBuilder;
+import org.ehcache.config.builders.ResourcePoolsBuilder;
+import org.ehcache.config.units.MemoryUnit;
+
+/**
+ * Created by achikin on 5/9/16.
+ */
+public class CachedRemoteStreamProvider implements RemoteStreamProvider {
+
+    private CacheManager cacheManager;
+
+    public CachedRemoteStreamProvider(int size) {
+        cacheManager = CacheManagerBuilder.newCacheManagerBuilder()
+                .withCache("preConfigured",
+                        CacheConfigurationBuilder.newCacheConfigurationBuilder(URL.class, AudioStreamCache.class,
+                                ResourcePoolsBuilder.newResourcePoolsBuilder().heap(size, MemoryUnit.MB))
+                                .build())
+                .build(true);
+    }
+
+    private Cache<URL, AudioStreamCache> getCache() {
+        return cacheManager.getCache("preConfigured", URL.class, AudioStreamCache.class);
+    }
+
+    public InputStream getStream(URL uri) throws IOException {
+        Cache<URL, AudioStreamCache> cache = getCache();
+
+        AudioStreamCache stream = cache.get(uri);
+        if (stream == null) {
+            stream = new AudioStreamCache(uri);
+            AudioStreamCache exists = cache.putIfAbsent(uri, stream);
+            if (exists != null) {
+                stream = exists;
+            }
+        }
+        return new ByteArrayInputStream(stream.getBytes());
+    }
+
+    private static class AudioStreamCache {
+
+        private URL uri;
```

Comment:
i know, again, since this class is part of a limited sized cache, the less the mem the object hold,the best. The "uri" tells how to getBytes, but getBytes its just a lazy-init of the object, which wont use uri anymore once the bytes are retrieved. Actually, you could split this method in two "retrieve" and "get", and call retrieve on "putifabsent == null condition" as suggested previously